### DISCUSSION TITLE
Do BayesianNetwork object support weighted edges?

### DISCUSSION THREAD

**Discussion by borjasor** [pgmpy/pgmpy#1759] [Q&A] [answered]
BayesianNetwork class inherits from DAG class which itself inherits from DiGraph class which itself inherits from Graph class which supports weighed edges. This is why BayesianNetwork has the inherited method 'add_weighted_edges_from', however this method does not seem to do anything on a BayesianNetwork model. We could edit the "weight" parameter on the edges by:

`model['U']['V']["weight"] = weight_value`

But this has no effect on Inference. If weighting is supported, could you please explain how to apply it correctly during inference?

Thank you

**Comment by ankurankan** [comment]
@borjasor You are correct, the weight attribute isn't used for Bayesian Networks.  However, we do use the weights for specifying the parameters of Structural Equation Models: https://github.com/pgmpy/pgmpy/blob/dev/pgmpy/models/SEM.py

Were you expecting something specific from the weight attribute in Bayesian Networks?

**Reply by borjasor** [reply]
The reasoning behind weights being applied to BayesianNetworks edges is explained in [this comment](https://github.com/pgmpy/pgmpy/discussions/1630#discussioncomment-9477416) and I assume in the original question in that same thread. I do not have experience with SEM models...but after some reading it may be that these models are the way to implement the example that I described on my [previous comment](https://github.com/pgmpy/pgmpy/discussions/1630#discussioncomment-9477416)?

**Reply by borjasor** [reply]
@ankurankan sorry to comeback to this but it would be really useful to know if there is a way of implementing this weighting edges or how SEMs could do something similar as you suggested. As example, we have a Naive Bayes model with one root node and multiple leave nodes. They have their own CPDs defined but how can i implement a case in which I want to bias the influence of one of the leaves respect to the other ones? that is where weights in edges could be used.

**Reply by ankurankan** [reply, marked as answer]
@borjasor Sorry, I somehow missed replying to your last comment in #1630. The CPD of the node already encodes the edge weight of all the incoming edges to it and you can manipulate it to bias the influence.  For example:
```python
In [1]: from pgmpy.models import BayesianNetwork

In [2]: model = BayesianNetwork([('C', 'F1'), ('C', 'F2')])

In [3]: model = model.get_random_cpds(n_states=3)

In [4]: print(model.get_cpds('F1'))
+-------+---------------------+---------------------+---------------------+
| C     | C(0)                | C(1)                | C(2)                |
+-------+---------------------+---------------------+---------------------+
| F1(0) | 0.17312388346017654 | 0.418840840272899   | 0.46341612555494727 |
+-------+---------------------+---------------------+---------------------+
| F1(1) | 0.3293053790790536  | 0.48799293592348897 | 0.03068287575211273 |
+-------+---------------------+---------------------+---------------------+
| F1(2) | 0.49757073746076985 | 0.09316622380361196 | 0.5059009986929401  |
+-------+---------------------+---------------------+---------------------+

In [5]: print(model.get_cpds('F2'))
+-------+---------------------+---------------------+--------------------+
| C     | C(0)                | C(1)                | C(2)               |
+-------+---------------------+---------------------+--------------------+
| F2(0) | 0.27726085902655373 | 0.4649305672035038  | 0.1449886731635488 |
+-------+---------------------+---------------------+--------------------+
| F2(1) | 0.6383373856244118  | 0.42931938631173594 | 0.4107716325243371 |
+-------+---------------------+---------------------+--------------------+
| F2(2) | 0.0844017553490343  | 0.10575004648476019 | 0.444239694312114  |
+-------+---------------------+---------------------+--------------------+

In [6]: df = model.simulate(1000)
Generating for node: C:   0%|                                           | 0/3 [00:00<?, ?it/s]WARNING:pgmpy:Probability values don't exactly sum to 1. Differ by: 2.220446049250313e-16. Adjusting values.
Generating for node: F2:   0%|                                          | 0/3 [00:00<?, ?it/s]WARNING:pgmpy:Probability values don't exactly sum to 1. Differ by: 1.1102230246251565e-16. Adjusting values.
Generating for node: F2: 100%|█████████████████████████████████| 3/3 [00:00<00:00, 590.28it/s]

In [7]: from pgmpy.estimators.CITests import pillai_trace

In [8]: pillai_trace('C', 'F1', [], df, boolean=False)[0]  # This is an effect size measure. Think of it as the strength of relationship between C and F1.
Out[8]: np.float64(0.3064411166836729)

In [9]: pillai_trace('C', 'F2', [], df, boolean=False)[0]
Out[9]: np.float64(0.1776121141584744)

# Modify the CPDs to reduce the strength of F1 and increase the strength of F2.

In [10]: model.get_cpds('F1').values[:, 0] = [0.33, 0.33, 0.34] # Making the first column probabilities equal implies that the state of variable F1 doesn't give us any information about the first state of C.

In [11]: model.get_cpds('F2').values[:, 0] = [0.1, 0.1, 0.8]

In [12]: df_mod = model.simulate(1000)
Generating for node: C:   0%|                                           | 0/3 [00:00<?, ?it/s]WARNING:pgmpy:Probability values don't exactly sum to 1. Differ by: 2.220446049250313e-16. Adjusting values.
Generating for node: F2: 100%|█████████████████████████████████| 3/3 [00:00<00:00, 400.07it/s]

In [16]: model.get_cpds('F1').values
Out[16]: 
array([[0.33      , 0.41884084, 0.46341613],
       [0.33      , 0.48799294, 0.03068288],
       [0.34      , 0.09316622, 0.505901  ]])

In [17]: model.get_cpds('F2').values
Out[17]: 
array([[0.1       , 0.46493057, 0.14498867],
       [0.1       , 0.42931939, 0.41077163],
       [0.8       , 0.10575005, 0.44423969]])

In [18]: pillai_trace('C', 'F1', [], df_mod, boolean=False)[0] # This value is lower than before modification.
Out[18]: np.float64(0.24770571158024296)

In [19]: pillai_trace('C', 'F2', [], df_mod, boolean=False)[0] # This value is higher.
Out[19]: np.float64(0.3444766753011126)
```
But as you can see interpreting these modifications is not the easiest. I think this is a general problem when we have categorical variables. In the case, when we have continuous variables, this is much easier as we have a single value to manipulate.
```python
In [20]: from pgmpy.models import LinearGaussianBayesianNetwork

In [21]: model = LinearGaussianBayesianNetwork([('C', 'F1'), ('C', 'F2')])

In [24]: model.add_cpds(*model.get_random_cpds())

In [25]: model.get_cpds()
Out[25]: 
[<LinearGaussianCPD: P(C) = N(0.642; 0.382) at 0x7dd92d661880,
 <LinearGaussianCPD: P(F1 | C) = N(-0.601*C + 0.078; 0.976) at 0x7dd92c1820c0,
 <LinearGaussianCPD: P(F2 | C) = N(-0.038*C + 0.272; 1.227) at 0x7dd92c183e30]

In [26]: model.to_joint_gaussian()[1]
Out[26]: 
 array([[ 0.38172424, -0.22940008, -0.01453796],
        [-0.22940008,  1.1141308 ,  0.0087367 ],
        [-0.01453796,  0.0087367 ,  1.22793535]])

In [32]: cpd_mod = LinearGaussianCPD('F2', [0.272, 1], 1.227, ['C']) # Increase the strength between F2 and C.

In [33]: model.add_cpds(cpd_mod)
WARNING:pgmpy:Replacing existing CPD for F2

In [34]: model.to_joint_gaussian()[1] # The covariance between F2 and C has changed from -0.0145 to 0.3817
Out[34]: 
 array([[ 0.38172424, -0.22940008,  0.38172424],
        [-0.22940008,  1.1141308 , -0.22940008],
        [ 0.38172424, -0.22940008,  1.60872424]])
```
In this continuous case, you can see that the covariance between F2 and C has been changed easily by manipulating a single parameter.

**Reply by borjasor** [reply]
Thank you so much for your detailed reply @ankurankan. I understand that your answer is a workaround to the fact that edge weights are not implemented? Your example shows how a lower weight could be implemented by changing the original CPD of F1 increasing uncertainty of each state of C. Of course this would have to be done for all states of C in a way that is common to all of them (as it would be the case of a single weight on the edge between C and F1). Could this be done by adding intermediate nodes (between F1 and C for instance) where its CPD implements the effect of increasing uncertainty of states of C? as you said before, this is similar to the concept of virtual evidence but of course this would not be implemented by a binary node child like in virtual evidences. I am not sure what this intermediate node would be.

**Reply by ankurankan** [reply]
@borjasor 
> I understand that your answer is a workaround to the fact that edge weights are not implemented?

Not exactly. I meant that there is no way to define an edge weight when discrete variables are involved (to the best of my knowledge; would be curious to see if you know of any approaches?). The Pillai's trace based effect size that I have shown in the code is the closest approach to assigning an edge weight for discrete variables, however, it can not be incorporated into downstream tasks such as inference. The only ways to modify the influence of specific variables (that I can think of) is to modify the CPDs as I have shown or modify the regression coefficients (which would also be a matrix of values) if using multinomial regression for parameterization.

> Could this be done by adding intermediate nodes (between F1 and C for instance) where its CPD implements the effect of increasing uncertainty of states of C? 

I think we would again be in a similar situation as we would need to define the CPD for this intermediate node and then to adjust the influence of the feature, somehow modify the CPDs again (similar to my code example).

> this is similar to the concept of virtual evidence but of course this would not be implemented by a binary node child like in virtual evidences. 

It could be possible using virtual evidence. The idea would be to define a virtual evidence for the feature and try to define the virtual evidence probabilities in such a way that it manipulates the influence of the feature. But I am not able to think of a way to define such virtual evidences.
